File: strands-ts/src/middleware/registry.ts
import type {
  MiddlewareStage,
  MiddlewareHandler,
  MiddlewareNext,
  MiddlewarePhaseKind,
  MiddlewareInputPhase,
  MiddlewareOutputPhase,
  MiddlewareInputHandler,
  MiddlewareOutputHandler,
} from './types.js'

/** Internal tagged handler with phase metadata for compose ordering. */
interface TaggedHandler<TContext, TEvent, TResult> {
  phase: MiddlewarePhaseKind
  handler: MiddlewareHandler<TContext, TEvent, TResult>
}

/** Phase compose order: input (outermost) → output → around (innermost, closest to terminal). */
const PHASE_ORDER: Record<MiddlewarePhaseKind, number> = { input: 0, output: 1, around: 2 }

/**
 * Registry that stores middleware handlers keyed by stage tokens
 * and composes them into execution chains with phase ordering.
 */
export class MiddlewareRegistry {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private readonly _handlers: Map<MiddlewareStage<any, any, any>, TaggedHandler<any, any, any>[]>

  constructor() {
    this._handlers = new Map()
  }

  /**
   * Register a middleware handler for a given stage (Around phase).
   * Handlers are stored in registration order within their phase.
   *
   * @param stage - The stage token to register the handler for
   * @param handler - The middleware handler function
   */
  add<TContext, TEvent, TResult>(
    stage: MiddlewareStage<TContext, TEvent, TResult>,
    handler: MiddlewareHandler<TContext, TEvent, TResult>
  ): void {
    const handlers = this._handlers.get(stage) ?? []
    handlers.push({ phase: 'around', handler })
    this._handlers.set(stage, handlers)
  }

  /**
   * Register an Input phase handler. Transforms context before the Around chain runs.
   * Multiple Input handlers compose in registration order (each sees the previous handler's output).
   * Returns the adapted handler for cleanup purposes.
   */
  addInput<TContext, TEvent, TResult>(
    phase: MiddlewareInputPhase<TContext, TEvent, TResult>,
    handler: MiddlewareInputHandler<TContext>
  ): MiddlewareHandler<TContext, TEvent, TResult> {
    const stage = phase._stage
    const handlers = this._handlers.get(stage) ?? []
    const adapted: MiddlewareHandler<TContext, TEvent, TResult> = async function* (context, next) {
      const transformed = await handler(context)
      return yield* next(transformed)
    }
    handlers.push({ phase: 'input', handler: adapted })
    this._handlers.set(stage, handlers)
    return adapted
  }

  /**
   * Register an Output phase handler. Transforms result after the Around chain completes.
   * Multiple Output handlers compose in registration order (each sees the previous handler's output).
   * Returns the adapted handler for cleanup purposes.
   */
  addOutput<TContext, TEvent, TResult>(
    phase: MiddlewareOutputPhase<TContext, TEvent, TResult>,
    handler: MiddlewareOutputHandler<TResult>
  ): MiddlewareHandler<TContext, TEvent, TResult> {
    const stage = phase._stage
    const handlers = this._handlers.get(stage) ?? []
    const adapted: MiddlewareHandler<TContext, TEvent, TResult> = async function* (context, next) {
      const result = yield* next(context)
      return await handler(result)
    }
    handlers.push({ phase: 'output', handler: adapted })
    this._handlers.set(stage, handlers)
    return adapted
  }

  /**
   * Remove a previously registered middleware handler for a given stage.
   * Removes the first occurrence of the handler (by reference equality on the adapted handler).
   *
   * @param stage - The stage token to remove the handler from
   * @param handler - The middleware handler function to remove
   */
  remove<TContext, TEvent, TResult>(
    stage: MiddlewareStage<TContext, TEvent, TResult>,
    handler: MiddlewareHandler<TContext, TEvent, TResult>
  ): void {
    const handlers = this._handlers.get(stage)
    if (!handlers) return
    const idx = handlers.findIndex((h) => h.handler === handler)
    if (idx !== -1) handlers.splice(idx, 1)
  }

  /**
   * Compose all registered handlers for a stage into a single middleware chain.
   * Handlers are ordered by phase (input → output → around), then by registration order within phase.
   * First in the composed chain = outermost.
   *
   * @param stage - The stage token to compose handlers for
   * @param terminal - The innermost function that performs actual stage execution
   * @returns A single function representing the full middleware chain
   */
  compose<TContext, TEvent, TResult>(
    stage: MiddlewareStage<TContext, TEvent, TResult>,
    terminal: MiddlewareNext<TContext, TEvent, TResult>
  ): MiddlewareNext<TContext, TEvent, TResult> {
    const tagged = (this._handlers.get(stage) ?? []) as TaggedHandler<TContext, TEvent, TResult>[]

    // Stable sort by phase: input first, output second, around last (closest to terminal)
    const sorted = [...tagged].sort((a, b) => PHASE_ORDER[a.phase] - PHASE_ORDER[b.phase])

    let current: MiddlewareNext<TContext, TEvent, TResult> = terminal
    for (let i = sorted.length - 1; i >= 0; i--) {
      const handler = sorted[i]!.handler
      const next = current
      current = (context: TContext): AsyncGenerator<TEvent, TResult, undefined> => handler(context, next)
    }

    return current
  }

  /**
   * Compose and invoke the middleware chain for a stage in one call.
   * Equivalent to `compose(stage, terminal)(context)` but reads more clearly at call sites.
   *
   * @param stage - The stage token to invoke
   * @param context - The context to pass into the chain
   * @param terminal - The innermost function that performs actual stage execution
   * @returns An async generator yielding events and returning the stage result
   */
  invoke<TContext, TEvent, TResult>(
    stage: MiddlewareStage<TContext, TEvent, TResult>,
    context: TContext,
    terminal: MiddlewareNext<TContext, TEvent, TResult>
  ): AsyncGenerator<TEvent, TResult, undefined> {
    const chain = this.compose(stage, terminal)
    return chain(context)
  }
}

File: strands-py/src/strands/_middleware/registry.py
"""Middleware registry for composing handler chains."""

from __future__ import annotations

import inspect
from collections.abc import AsyncGenerator
from dataclasses import dataclass
from typing import Any

from .types import (
    MiddlewareHandler,
    MiddlewareInputHandler,
    MiddlewareInputPhase,
    MiddlewareNext,
    MiddlewareOutputHandler,
    MiddlewareOutputPhase,
    MiddlewareResult,
    MiddlewareStage,
    MiddlewareWrapPhase,
)

_PHASE_ORDER: dict[str, int] = {"input": 0, "output": 1, "wrap": 2}


@dataclass
class _TaggedHandler:
    phase: str
    handler: MiddlewareHandler


class MiddlewareRegistry:
    """Registry that stores middleware handlers keyed by stage tokens and composes them into chains."""

    def __init__(self) -> None:
        self._handlers: dict[MiddlewareStage[Any, Any, Any], list[_TaggedHandler]] = {}

    def add_middleware(
        self,
        stage_or_phase: (
            MiddlewareStage[Any, Any, Any]
            | MiddlewareInputPhase[Any, Any, Any]
            | MiddlewareWrapPhase[Any, Any, Any]
            | MiddlewareOutputPhase[Any, Any, Any]
        ),
        handler: Any,
    ) -> None:
        """Register middleware for a stage or phase sub-token."""
        if isinstance(stage_or_phase, MiddlewareInputPhase):
            self._add_input(stage_or_phase, handler)
        elif isinstance(stage_or_phase, MiddlewareOutputPhase):
            self._add_output(stage_or_phase, handler)
        elif isinstance(stage_or_phase, MiddlewareWrapPhase):
            self._add_wrap(stage_or_phase._stage, handler)
        else:
            self._add_wrap(stage_or_phase, handler)

    def _add_wrap(self, stage: MiddlewareStage[Any, Any, Any], handler: MiddlewareHandler) -> None:
        handlers = self._handlers.setdefault(stage, [])
        handlers.append(_TaggedHandler(phase="wrap", handler=handler))

    def _add_input(self, phase: MiddlewareInputPhase[Any, Any, Any], handler: MiddlewareInputHandler) -> None:
        stage = phase._stage

        async def adapted(context: Any, next_fn: MiddlewareNext) -> AsyncGenerator[Any, None]:
            transformed = handler(context)
            if inspect.isawaitable(transformed):
                transformed = await transformed
            async for event in next_fn(transformed):
                yield event

        handlers = self._handlers.setdefault(stage, [])
        handlers.append(_TaggedHandler(phase="input", handler=adapted))

    def _add_output(self, phase: MiddlewareOutputPhase[Any, Any, Any], handler: MiddlewareOutputHandler) -> None:
        stage = phase._stage

        async def adapted(context: Any, next_fn: MiddlewareNext) -> AsyncGenerator[Any, None]:
            async for event in next_fn(context):
                if isinstance(event, MiddlewareResult):
                    transformed = handler(event.value)
                    if inspect.isawaitable(transformed):
                        transformed = await transformed
                    yield MiddlewareResult(transformed)
                else:
                    yield event

        handlers = self._handlers.setdefault(stage, [])
        handlers.append(_TaggedHandler(phase="output", handler=adapted))

    def compose(self, stage: MiddlewareStage[Any, Any, Any], terminal: MiddlewareNext) -> MiddlewareNext:
        """Compose all registered handlers for a stage into a single chain.

        Returns the terminal directly if no handlers are registered (zero overhead fast path).
        """
        tagged = self._handlers.get(stage)
        if not tagged:
            return terminal

        sorted_handlers = sorted(tagged, key=lambda t: _PHASE_ORDER[t.phase])

        current: MiddlewareNext = terminal
        for i in range(len(sorted_handlers) - 1, -1, -1):
            handler = sorted_handlers[i].handler
            next_fn = current

            def _make_layer(h: MiddlewareHandler, nf: MiddlewareNext) -> MiddlewareNext:
                async def layer(ctx: Any) -> AsyncGenerator[Any, None]:
                    inner_gens: list[AsyncGenerator[Any, None]] = []

                    def tracking_next(c: Any) -> AsyncGenerator[Any, None]:
                        gen = nf(c)
                        inner_gens.append(gen)
                        return gen

                    handler_gen = h(ctx, tracking_next)
                    try:
                        async for event in handler_gen:
                            yield event
                    finally:
                        await handler_gen.aclose()
                        for gen in inner_gens:
                            await gen.aclose()

                return layer

            current = _make_layer(handler, next_fn)

        return current

    async def invoke(
        self,
        stage: MiddlewareStage[Any, Any, Any],
        context: Any,
        terminal: MiddlewareNext,
    ) -> AsyncGenerator[Any, None]:
        """Compose and invoke the middleware chain for a stage."""
        chain = self.compose(stage, terminal)
        gen = chain(context)
        try:
            async for event in gen:
                yield event
        finally:
            await gen.aclose()
